service/history/queues/rescheduler.go

```diff
@@ -75,71 +82,135 @@ type (
+
 		pq: collection.NewPriorityQueue((func(this rescheduledExecuable, that rescheduledExecuable) bool {
 			return this.rescheduleTime.Before(that.rescheduleTime)
 		})),
 	}
 }
 
+func (r *reschedulerImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	r.shutdownWG.Add(1)
+	go r.rescheduleLoop()
+
+	r.logger.Info("Task rescheduler started.", tag.LifeCycleStarted)
+}
+
+func (r *reschedulerImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(r.shutdownCh)
+	r.timerGate.Close()
+
+	if success := common.AwaitWaitGroup(&r.shutdownWG, time.Minute); !success {
+		r.logger.Warn("Task rescheduler timedout on shutdown.", tag.LifeCycleStopTimedout)
+	}
+
+	r.logger.Info("Task rescheduler stopped.", tag.LifeCycleStopped)
+}
+
 func (r *reschedulerImpl) Add(
 	executable Executable,
-	backoff time.Duration,
+	rescheduleTime time.Time,
 ) {
 	r.Lock()
-	defer r.Unlock()
-
 	r.pq.Add(rescheduledExecuable{
 		executable:     executable,
-		rescheduleTime: r.timeSource.Now().Add(backoff),
+		rescheduleTime: rescheduleTime,
 	})
+	r.Unlock()
+
+	r.timerGate.Update(rescheduleTime)
+
+	if r.isStopped() {
+		r.drain()
+	}
 }
 
-func (r *reschedulerImpl) Reschedule(
-	targetRescheduleSize int,
-) {
+func (r *reschedulerImpl) Len() int {
 	r.Lock()
 	defer r.Unlock()
 
-	r.metricProvider.Histogram(TaskReschedulerPendingTasks, metrics.Dimensionless).Record(int64(r.pq.Len()))
+	return r.pq.Len()
+}
 
-	if targetRescheduleSize == 0 {
-		targetRescheduleSize = r.pq.Len()
+func (r *reschedulerImpl) rescheduleLoop() {
+	defer r.shutdownWG.Done()
+
+	for {
+		select {
+		case <-r.shutdownCh:
+			r.drain()
+			return
+		case <-r.timerGate.FireChan():
+			r.reschedule()
+		}
 	}
 
+}
+
+func (r *reschedulerImpl) reschedule() {
+	r.Lock()
+	defer r.Unlock()
+
+	r.metricsHandler.Histogram(TaskReschedulerPendingTasks, metrics.Dimensionless).Record(int64(r.pq.Len()))
+
 	var failToSubmit []rescheduledExecuable
-	numRescheduled := 0
-	for !r.pq.IsEmpty() && numRescheduled < targetRescheduleSize {
+	for !r.pq.IsEmpty() {
 		if r.timeSource.Now().Before(r.pq.Peek().rescheduleTime) {
 			break
 		}
 
 		rescheduled := r.pq.Remove()
-		submitted, err := r.scheduler.TrySubmit(rescheduled.executable)
+		executable := rescheduled.executable
+		submitted, err := r.scheduler.TrySubmit(executable)
 		if err != nil {
-			rescheduled.executable.Logger().Error("Failed to reschedule task", tag.Error(err))
+			executable.Logger().Error("Failed to reschedule task", tag.Error(err))
 		}
 
 		if !submitted {
+			rescheduled.rescheduleTime.Add(rescheduleFailureBackoff)
 			failToSubmit = append(failToSubmit, rescheduled)
-		} else {
-			numRescheduled++
 		}
 	}
 
 	for _, rescheduled := range failToSubmit {
 		r.pq.Add(rescheduled)
 	}
+
+	if !r.pq.IsEmpty() {
+		r.timerGate.Update(r.pq.Peek().rescheduleTime)
```

Comment:
do you need to update if it break from L 180?

Reply:
yes, otherwise the next round of rescheduling won't be triggered. This behavior is asserted by the first test in the test suite (`TestStartStop`).